### PR TITLE
New cypher api changes / improvements

### PR DESCRIPTION
**Pull request by cheerfulstoic**

### CODE REVIEW

**Comment by cheerfulstoic** [comment]
I wanted to give an overview of what's happening.  Things on here aren't neccesarily done, but I wanted to talk about what I've been hammering away at.  This is a summary of things that I've been working on it both gems.

- ActiveNode / ActiveRel now use the new `Neo4j::Core::CypherSession` API.  When all is said and done I'd love to run benchmarks comparing before/after

- Transactions have been refactored.  I think that it's cleaner now and there is a better separation of concerns.  It fits better with standardizing transactions between the old session API and the new one.  Also, instead of a @pushed_nested variable there is now a transaction stack (Array) to track all of the transactions which have been piled up so far

- No longer "current" transaction.  To get a transaction, you must look it up with a session object.  The specs use a current transaction, though, for convenience.

- To run a transaction you need to specify a session.  Can use either Transaction.new(run_in_tx, session) or Transaction.new(session, run_in_tx).  If you only specify run_in_tx, the `current!` method is called for backwards compatibility

- "Current" session is there, though as much as possible I've tried to make it so that things are done by a session that the user creates (I also read this blog post by tenderlove recently and I think he had similar thoughts: http://tenderlovemaking.com/2011/10/20/connection-management-in-activerecord.html)

- ActiveNode/ActiveRel now have a `ActiveBase` module, which right now is just a way to hold the current session for queries to from ActiveNode/ActiveRel.  The railtie sets this up.  It also uses the `PerThreadRegistry` so it should be thread safe.  For configurations like puma/unicorn I think we'll need to write / test some template `on_worker_boot` code for forks/threads.  Maybe pooling?  Later!

- New Session API will let you choose how query results are wrapped (Ruby data, Node/Relationship/Path entities, or custom wrapping like ActiveNode/ActiveRel).  Can set a global configuration variable and overwrite per-query (default is to wrap with Node/Relationship/Path classes)

- Work in progress: I've been exploring the idea of having a second session in the background which is responsible for all database schema changes.  This means that the indexes/constraints defined in the model won't get mixed up with data changes in transaction.  This is also happening in a thread because if both a data change and a transaction are happening at the same time then things will lock up.  If we're waiting on an HTTP request, though, the thread waiting won't block the other thread.  This is *probably* a good idea, but feedback/discussion is welcome!

**Comment by cheerfulstoic** [comment]
Oh, and more:

 - To specify session configuration the preferred is now `http` and `embedded` (`server_db` and `embedded_db` still supported).  Eventually this will also include `bolt` (`server_db` isn't so good any more because `http` and `bolt` are both server)
 - I removed "named" sessions. I don't think that anybody is using that

**Comment by cheerfulstoic** [comment]
To see an example of how you can get different levels of wrapping from a query, see this commit:

https://github.com/neo4jrb/neo4j-core/commit/8ca81846c8e74e511a2064a287e41cd21c2eb86a

Need to test the global config as well

File: lib/neo4j/core/cypher_session/adaptors.rb [hunk 1]
```diff
@@ -52,68 +57,71 @@ def query
             end
           end
 
-          def query(*args)
-            queries { append(*args) }[0]
+          def query(session, *args)
+            options = (args.size == 3 || (args[0].is_a?(::Neo4j::Core::Query) && args.size == 2)) ? args.pop : {}
```
Comment (subvertallchris) [inline lib/neo4j/core/cypher_session/adaptors.rb]: Why does size matter?

:trollface: 

But really, is there another way to express this that can help the reader understand what this means?

Comment (cheerfulstoic) [inline lib/neo4j/core/cypher_session/adaptors.rb]: Yeah, good point ;)  I'll clean this up

**Comment by cheerfulstoic** [comment]
Ahhh, that feels good ;)

File: lib/neo4j/core/cypher_session/adaptors.rb (hunk 1 above)
Comment (cheerfulstoic) [inline lib/neo4j/core/cypher_session/adaptors.rb]: Fixed! https://github.com/neo4jrb/neo4j-core/blob/master/lib/neo4j/core/cypher_session/adaptors.rb#L73

**Pull request merged**